In [ ]:
import pandas as pd
from sklearn.naive_bayes import GaussianNB
import numpy as np 
import matplotlib.pyplot as plt
import math as mt
import io 

#Importando os datasets - Treinamento
Os datasets estão na pasta info e são especificados como arquivos do tipo csv. Avila-tr - > training data set Avita-ts - > test dataset

O primeiro objetivo é importar os dois datasets e dropar a coluna de labels

In [ ]:
!wget -q https://raw.githubusercontent.com/pvesouza/datasets/main/avila-tr.csv
!wget -q https://raw.githubusercontent.com/pvesouza/datasets/main/avila-ts.csv

In [ ]:
avilaDatasetTr = pd.read_csv('/content/avila-tr.csv')
print(avilaDatasetTr.head(10))
print(avilaDatasetTr.tail(10))
avilaDatasetTr.columns

         F1        F2        F3        F4        F5        F6        F7  \
0  0.266074 -0.165620  0.320980  0.483299  0.172340  0.273364  0.371178   
1  0.130292  0.870736 -3.210528  0.062493  0.261718  1.436060  1.465940   
2 -0.116585  0.069915  0.068476 -0.783147  0.261718  0.439463 -0.081827   
3  0.031541  0.297600 -3.210528 -0.583590 -0.721442 -0.307984  0.710932   
4  0.229043  0.807926 -0.052442  0.082634  0.261718  0.148790  0.635431   
5  0.117948 -0.220579 -3.210528 -1.623238  0.261718 -0.349509  0.257927   
6  0.389513 -0.220579 -3.210528 -2.624155  0.261718 -0.764757  0.484429   
7  0.019197 -0.040001  0.288973 -0.042597  0.261718 -1.013906  0.069175   
8  0.500607  0.140576  0.388552 -0.637358  0.261718 -0.681707  0.295677   
9 -0.252367  0.069915  0.246296  0.523550  0.261718 -1.221530  0.899684   

         F8        F9       F10 F11  
0  0.929823  0.251173  0.159345   A  
1  0.636203  0.282354  0.515587   A  
2 -0.888236 -0.123005  0.582939   A  
3  1.051693  0.594169 

Index(['F1', 'F2', 'F3', 'F4', 'F5', 'F6', 'F7', 'F8', 'F9', 'F10', 'F11'], dtype='object')

In [ ]:
avilaDatasetTs = pd.read_csv('/content/avila-ts.csv')
print(avilaDatasetTs.head(5))
print(avilaDatasetTs.shape)

         F1        F2        F3        F4        F5        F6        F7  \
0 -3.498799  0.250492  0.232070  1.224178 -4.922215  1.145386  0.182426   
1  0.204355 -0.354049  0.320980  0.410166 -0.989576 -2.218127  0.220177   
2  0.759828 -1.304042 -0.023991 -0.973663 -0.006417 -0.349509 -0.421580   
3 -0.005490  0.360409  0.281860 -0.213479 -1.168333 -1.013906 -0.346080   
4  0.080916  0.101320  0.104040  0.140490  0.261718  0.480988  0.710932   

         F8        F9       F10 F11  
0 -0.165983 -0.123005  1.087144   W  
1  0.181844  2.090879 -2.009758   A  
2 -0.450127  0.469443  0.060952   I  
3  1.176165  0.968347 -0.627999   E  
4 -0.253430 -0.497183  0.155681   A  
(10437, 11)


# Juntando os datasets

In [ ]:
frames = [avilaDatasetTs, avilaDatasetTr]
joinDataset = pd.concat(frames, ignore_index=True)

print(joinDataset.head(10))
print(joinDataset.tail(10))

print("New dataset shape: ", joinDataset.shape)

         F1        F2        F3        F4        F5        F6        F7  \
0 -3.498799  0.250492  0.232070  1.224178 -4.922215  1.145386  0.182426   
1  0.204355 -0.354049  0.320980  0.410166 -0.989576 -2.218127  0.220177   
2  0.759828 -1.304042 -0.023991 -0.973663 -0.006417 -0.349509 -0.421580   
3 -0.005490  0.360409  0.281860 -0.213479 -1.168333 -1.013906 -0.346080   
4  0.080916  0.101320  0.104040  0.140490  0.261718  0.480988  0.710932   
5  0.068573 -0.181323 -3.210528 -0.294311 -1.168333  0.356414 -0.006326   
6 -0.301743 -0.314793  0.399221  0.770520  0.708609  0.564038 -1.403091   
7  0.031541 -0.118513  0.374326 -0.066706  0.261718  0.605563  0.559930   
8 -0.091897 -0.118513  0.189393  1.280303  0.261718  0.314889  0.069175   
9  0.377169  0.014957  0.381439  0.292753  0.261718 -0.307984  0.522180   

         F8        F9       F10 F11  
0 -0.165983 -0.123005  1.087144   W  
1  0.181844  2.090879 -2.009758   A  
2 -0.450127  0.469443  0.060952   I  
3  1.176165  0.968347 

# Procurando por linhas NaN

In [ ]:
# Procurando por linhas com linhas com NaN
dfValues = joinDataset.values
lenght = len(dfValues)
columns = dfValues[0].size

print(lenght)
print(columns)

for i in range(0, columns - 1):
    print(dfValues[0][i])
    print(type(dfValues[0][i]))

for r in range(0,lenght):
    for c in range(0, columns - 1):
        value = dfValues[r][c]
        if mt.isnan(value):
            print(r)

20867
11
-3.498799
<class 'float'>
0.250492
<class 'float'>
0.23207
<class 'float'>
1.224178
<class 'float'>
-4.922215
<class 'float'>
1.145386
<class 'float'>
0.182426
<class 'float'>
-0.165983
<class 'float'>
-0.123005
<class 'float'>
1.087144
<class 'float'>


# Separando os conjuntos entre teste e treinamento


In [ ]:
from pandas.core.common import random_state
from sklearn.model_selection import train_test_split

data = joinDataset.to_numpy()
nRow, nCol =  data.shape
print(nRow)
print(nCol)

Y = data[:,-1]
X = data[:, 0 : nCol - 1]
#Separando entre dados de treinamento e teste


X_train, X_test, Y_train, Y_test = train_test_split(X, Y, train_size = 0.7, random_state = 10)

print(np.min(X_train[:,3]))

20867
11
-7.0


#Separando um conjunto de validação a partir do conjunto de teste

In [ ]:
print(pd.DataFrame(Y_train).value_counts())
print(pd.DataFrame(X_train).shape)

A    6005
F    2746
E    1537
I    1174
H     728
X     718
G     621
D     474
Y     375
C     157
W      64
B       7
dtype: int64
(14606, 10)


#Implementando o modelo Nayve Bayes Gaussiano

In [ ]:
from scipy.stats import multivariate_normal

classes = joinDataset['F11'].unique()

#Matriz para armazenar as probabilidades de cada classe
P = pd.DataFrame(data=np.zeros((X_train.shape[0], len(classes))), columns=classes)
#Fração de elementos de cada classes
Pc = np.zeros(len(classes))

#For para iterar sobre cada classe i
for i in range(0, len(classes)):
  elements = tuple(np.where(Y_train == classes[i]))
  Pc[i] = len(elements) / len(Y_train)                                  #Probabilidades a priori
  Z = X_train[elements,:][0]
  mean = np.mean(Z, axis=0)                                             #Vetor de médias de cada class
  covar = np.cov(np.transpose(Z.astype(float)))                         #Covariância dos elementos xi
  print(classes[i])
  for j in range(0, X_test.shape[0]):
    x = X_test[j,:]
    pj = multivariate_normal.pdf(x, mean=mean, cov=covar, allow_singular=True)
    P[classes[i]][j] = Pc[i] * pj

W
A
I
E
F
Y
H
D
G
X
C
B


In [ ]:
print(P)

                  W              A             I             E             F  \
0      0.000000e+00  2.266913e-265  0.000000e+00  0.000000e+00  0.000000e+00   
1      5.638253e-60   1.708393e-12  1.051834e-11  1.026929e-11  2.240110e-11   
2      3.542028e-39   5.558652e-12  1.052023e-10  4.000762e-12  1.448363e-10   
3      4.712343e-42   1.822474e-13  1.524768e-09  7.019596e-13  1.610175e-12   
4      4.363344e-11   1.844292e-12  2.359917e-13  1.404387e-12  1.810935e-11   
...             ...            ...           ...           ...           ...   
14601  0.000000e+00   0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
14602  0.000000e+00   0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
14603  0.000000e+00   0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
14604  0.000000e+00   0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
14605  0.000000e+00   0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   

                  Y             H      

In [ ]:
print(Y_test[0])

NameError: ignored

#Predição

In [ ]:
from numpy.lib.function_base import append
y_pred = []

for i in range(0, Y_test.shape[0]):
  c = np.argmax(np.array(P.iloc[[i]]))
  y_pred.append(classes[c])

#Passando para numpy
y_pred = np.array(y_pred, dtype=str)

In [ ]:
#Avaliando o preditor
from sklearn.metrics import accuracy_score
score = accuracy_score(y_pred, Y_test)

print("Score: [%.2f]" % (score * 100))

Score: [17.20]


# Implementando o K_Vizinhos

'\nmaior_acuracia = 0\ndist = 0\nviz = 0\nfor d in distancias:\n  for v in vizinhos:\n    knn = KNeighborsClassifier(n_neighbors = v, metric = d)\n    knn.fit(X_train, Y_train)\n    accuracy = knn.score(X_valid_train, Y_valid_train)\n    if maior_acuracia < accuracy:\n      maior_acuracia = accuracy\n      dist = d\n      viz = v\n\nprint(dist)\nprint(viz)\nprint(maior_acuracia)\n'

0.6833122044199889
{'n_neighbors': 16, 'metric': 'manhattan'}


0.8191183517010062


#Regressão Logística - um contra todos

Transformando a base de treinamento em 12 bases (classe de interese vs ¬classe de interesse)

array([[0.15498, -0.16562, -3.210528, ..., 0.340506, 0.157628, -0.415334],
       [0.327793, 0.50173, 0.050694, ..., 0.754385, 0.812439, -0.74098],
       [0.080916, 0.10132, 0.10404, ..., -0.327849, -0.247731, 0.564989],
       ...,
       [0.080916, 0.014957, 0.064919, ..., 0.264107, 0.282354, 0.574114],
       [0.216699, -0.291239, 0.349432, ..., 0.523361, 0.18881, -0.201646],
       [-0.437525, 0.423218, 0.388552, ..., 1.587403, 1.155435, 0.304429]],
      dtype=object)

array(['A', 'G', 'A', ..., 'A', 'A', 'E'], dtype=object)

In [ ]:
type(Y_train)

numpy.ndarray

In [ ]:
ova_databases = {}

In [ ]:
for scribe in np.unique(Y_train):
  temp = np.where(Y_train == scribe)  
  class_idx = 0
  new_y_train = np.ndarray(Y_train.shape, dtype=object)
  idx_cidx = []
  for idx, line in enumerate(X_train):
    if  class_idx < len(temp[0]) and temp[0][class_idx] == idx:
      idx_cidx.append(idx)
      new_y_train[idx] = Y_train[idx]
      class_idx += 1
    else:
      new_y_train[idx] = '!{}'.format(scribe)
  ova_databases[scribe] = new_y_train
    

In [ ]:
ova_databases

{'A': array(['A', '!A', 'A', ..., 'A', 'A', '!A'], dtype=object),
 'B': array(['!B', '!B', '!B', ..., '!B', '!B', '!B'], dtype=object),
 'C': array(['!C', '!C', '!C', ..., '!C', '!C', '!C'], dtype=object),
 'D': array(['!D', '!D', '!D', ..., '!D', '!D', '!D'], dtype=object),
 'E': array(['!E', '!E', '!E', ..., '!E', '!E', 'E'], dtype=object),
 'F': array(['!F', '!F', '!F', ..., '!F', '!F', '!F'], dtype=object),
 'G': array(['!G', 'G', '!G', ..., '!G', '!G', '!G'], dtype=object),
 'H': array(['!H', '!H', '!H', ..., '!H', '!H', '!H'], dtype=object),
 'I': array(['!I', '!I', '!I', ..., '!I', '!I', '!I'], dtype=object),
 'W': array(['!W', '!W', '!W', ..., '!W', '!W', '!W'], dtype=object),
 'X': array(['!X', '!X', '!X', ..., '!X', '!X', '!X'], dtype=object),
 'Y': array(['!Y', '!Y', '!Y', ..., '!Y', '!Y', '!Y'], dtype=object)}

In [ ]:
params = {'penalty': ['l1', 'l2', 'none'],
          'C': [0.5, 1.0, 1.5],
          'class_weight': ['balanced', None],
          'solver': ['newton-cg', 'lbfgs', 'liblinear'],
          'max_iter': [1000],}

In [ ]:
ova_estimators = {}
for scribe in np.unique(Y_train):
  print('searching scribe: {}'.format(scribe))
  reg_log = LogisticRegression()
  busca = RandomizedSearchCV(reg_log, params, n_iter=10, scoring='accuracy', cv=cv, random_state=10)
  result = busca.fit(X_train, ova_databases[scribe])
  ova_estimators[scribe] = result.best_estimator_

searching scribe: A
searching scribe: B
searching scribe: C
searching scribe: D
searching scribe: E
searching scribe: F
searching scribe: G
searching scribe: H
searching scribe: I


KeyboardInterrupt: ignored

In [ ]:
b1 = ova_estimators['A']

In [ ]:
ova_estimators

{'A': LogisticRegression(C=1.5, class_weight='balanced', max_iter=1000,
                    solver='newton-cg'),
 'B': LogisticRegression(C=1.5, class_weight='balanced', max_iter=1000, penalty='l1',
                    solver='liblinear'),
 'C': LogisticRegression(max_iter=1000),
 'D': LogisticRegression(max_iter=1000),
 'E': LogisticRegression(max_iter=1000),
 'F': LogisticRegression(max_iter=1000),
 'G': LogisticRegression(max_iter=1000),
 'H': LogisticRegression(max_iter=1000),
 'I': LogisticRegression(max_iter=1000),
 'W': LogisticRegression(max_iter=1000),
 'X': LogisticRegression(max_iter=1000),
 'Y': LogisticRegression(max_iter=1000)}

In [ ]:
ova_estimators['I'] = LogisticRegression(max_iter=1000).fit(X_train, ova_databases['I'])
ova_estimators['W'] = LogisticRegression(max_iter=1000).fit(X_train, ova_databases['W'])
ova_estimators['X'] = LogisticRegression(max_iter=1000).fit(X_train, ova_databases['X'])
ova_estimators['Y'] = LogisticRegression(max_iter=1000).fit(X_train, ova_databases['Y'])

In [ ]:
for scribe in ova_estimators.keys():
  temp = np.where(Y_test == scribe)  
  class_idx = 0
  new_y_test = np.ndarray(Y_test.shape, dtype=object)
  idx_cidx = []
  for idx, line in enumerate(X_test):
    if  class_idx < len(temp[0]) and temp[0][class_idx] == idx:
      idx_cidx.append(idx)
      new_y_test[idx] = Y_test[idx]
      class_idx += 1
    else:
      new_y_test[idx] = '!{}'.format(scribe)
  a_test_y = new_y_test
  print('scribe {} acc: {}'.format(scribe, ova_estimators[scribe].score(X_test, a_test_y)))

scribe A acc: 0.6818399616674653
scribe B acc: 0.9997604216578821
scribe C acc: 0.9913751796837565
scribe D acc: 0.9664590321034978
scribe E acc: 0.8852419741255391
scribe F acc: 0.8047436511739339
scribe G acc: 0.9551988500239579
scribe H acc: 0.9465740297077144
scribe I acc: 0.9750838524197413
scribe W acc: 0.9956875898418783
scribe X acc: 0.9753234307618591
scribe Y acc: 0.9712505989458553


In [ ]:
X_test[0]

array([-0.128929, -0.511072, 0.342319, 0.170775, 0.261718, 1.685209,
       -0.044076, -1096.0, -0.559546, 1.709756], dtype=object)

In [ ]:
ova_estimators['E'].predict([X_test[1]])

array(['!E'], dtype=object)

In [ ]:
ova_estimators['E'].predict_proba([X_test[1]])

array([[0.81330864, 0.18669136]])

In [ ]:
ova_estimators['D'].classes_

array(['!D', 'D'], dtype=object)

In [ ]:
Y_test[1]

'E'

In [ ]:
def ova_classify_one(estimators_dict, x):
  probabilities = []
  idx = 0
  for scribe in estimators_dict.keys():
    aux_prob = estimators_dict[scribe].predict_proba([x])
    aux_answer = estimators_dict[scribe].predict([x])
    #print('scribe: {}'.format(scribe))
    #print(aux_answer[0][0])
    idx+=1
    #print(idx, scribe)
    probabilities.append((scribe, aux_prob[0][1]))
    #print(probabilities)
    #print('entrou')
  
  sorted_probabilites = sorted(probabilities, key=lambda tup: tup[1], reverse=True)
  #print(sorted_probabilites)
  return sorted_probabilites[0]

def ova_classify_all(estimators_dict, X):
  predicted_y = []
  for x in X:
    predicted_y.append(ova_classify_one(estimators_dict, x))
  return predicted_y

pred_y = ova_classify_all(ova_estimators, X_test)

In [ ]:
pred_y

[('A', 0.6884275401087053),
 ('A', 0.30375032524266676),
 ('A', 0.600781627414917),
 ('I', 0.7547908727923179),
 ('A', 0.5638701986730569),
 ('A', 0.27289529964233644),
 ('A', 0.4256775264074503),
 ('I', 0.6226894593131106),
 ('A', 0.33217669110018944),
 ('A', 0.6179314921908232),
 ('A', 0.5423583211743783),
 ('F', 0.2433094734781482),
 ('A', 0.5829638480868351),
 ('A', 0.6232258456177857),
 ('A', 0.2686930781686193),
 ('A', 0.5307274052505351),
 ('A', 0.4990577776554794),
 ('A', 0.6111042623000866),
 ('A', 0.5285926771008088),
 ('X', 0.754773593175922),
 ('A', 0.35073337821931166),
 ('A', 0.47263409519483146),
 ('A', 0.6174956228514774),
 ('A', 0.4246747021479826),
 ('A', 0.6786483036381646),
 ('A', 0.9156370103031369),
 ('A', 0.8957772252233445),
 ('A', 0.3966353655535212),
 ('A', 0.8526227434659265),
 ('A', 0.6314638194999338),
 ('A', 0.33931187695685916),
 ('X', 0.8204333317723075),
 ('A', 0.42639983807198706),
 ('A', 0.24685359588105527),
 ('A', 0.48792720042307525),
 ('A', 0.8181

In [ ]:
list_y = []
for entry in pred_y:
  list_y.append(entry[0])

correct = 0
total = len(list_y)
for idx, pred in enumerate(list_y):
  if pred == Y_test[idx]:
    correct += 1

print('acc: {}%'.format(correct/total * 100))

acc: 50.143747005270725%


In [ ]:
from sklearn.metrics import classification_report

In [ ]:
classification_report(Y_test, list_y)

'              precision    recall  f1-score   support\n\n           A       0.46      0.98      0.63      1686\n           B       0.50      1.00      0.67         1\n           C       0.00      0.00      0.00        35\n           D       0.00      0.00      0.00       140\n           E       0.26      0.02      0.04       465\n           F       0.00      0.00      0.00       784\n           G       0.00      0.00      0.00       177\n           H       0.53      0.04      0.07       215\n           I       0.91      0.80      0.85       328\n           W       0.00      0.00      0.00        17\n           X       0.81      0.68      0.74       218\n           Y       0.42      0.05      0.08       108\n\n    accuracy                           0.50      4174\n   macro avg       0.32      0.30      0.26      4174\nweighted avg       0.37      0.50      0.37      4174\n'

              precision    recall  f1-score   support

           A       0.46      0.98      0.63      1686
           B       0.50      1.00      0.67         1
           C       0.00      0.00      0.00        35
           D       0.00      0.00      0.00       140
           E       0.26      0.02      0.04       465
           F       0.00      0.00      0.00       784
           G       0.00      0.00      0.00       177
           H       0.53      0.04      0.07       215
           I       0.91      0.80      0.85       328
           W       0.00      0.00      0.00        17
           X       0.81      0.68      0.74       218
           Y       0.42      0.05      0.08       108

    accuracy                           0.50      4174
   macro avg       0.32      0.30      0.26      4174
weighted avg       0.37      0.50      0.37      4174

In [ ]:
b1.score(X_test, a_test_y)

0.6818399616674653

In [ ]:
reg_log_ova = OneVsRestClassifier(LogisticRegression(max_iter=1000, random_state=10))
reg_log_ova.fit(X_train, Y_train)


OneVsRestClassifier(estimator=LogisticRegression(max_iter=1000,
                                                 random_state=10))

In [ ]:
reg_log_ova.score(X_test, Y_test)

0.5069477719214183

Observando a frequencia de cada classe

In [ ]:
df_X = pd.DataFrame(X_train)
df_Y = pd.DataFrame(Y_train)

frames = [df_X, df_Y]
df_Data = pd.concat(frames, axis=1, ignore_index = True)
df_Data.columns = joinDataset.columns

classes_frequency = df_Data.F11.value_counts()
classes = df_Data.F11.unique()

print(classes_frequency)

# Numero de elementos de cada classe
N_A = classes_frequency[0]
N_F = classes_frequency[1]
N_E = classes_frequency[2]
N_I = classes_frequency[3]
N_X = classes_frequency[4]
N_H = classes_frequency[5]
N_G = classes_frequency[6]
N_D = classes_frequency[7]
N_Y = classes_frequency[8]
N_C = classes_frequency[9]
N_W = classes_frequency[10]
N_B = classes_frequency[11]

frequencias = [N_A, N_F, N_E, N_I, N_X, N_H, N_G, N_D, N_Y, N_C, N_W, N_B]

#Recuperando os dataframes de cada classe

dfmask = df_Data['F11'] == 'A'
df_A = df_Data[dfmask]
dfmask = df_Data['F11'] == 'F'
df_F = df_Data[dfmask]
dfmask = df_Data['F11'] == 'E'
df_E = df_Data[dfmask]
dfmask = df_Data['F11'] == 'I'
df_I = df_Data[dfmask]
dfmask = df_Data['F11'] == 'X'
df_X = df_Data[dfmask]
dfmask = df_Data['F11'] == 'H'
df_H = df_Data[dfmask]
dfmask = df_Data['F11'] == 'G'
df_G = df_Data[dfmask]
dfmask = df_Data['F11'] == 'D'
df_D = df_Data[dfmask]
dfmask = df_Data['F11'] == 'Y'
df_Y = df_Data[dfmask]
dfmask = df_Data['F11'] == 'C'
df_C = df_Data[dfmask]
dfmask = df_Data['F11'] == 'W'
df_W = df_Data[dfmask]
dfmask = df_Data['F11'] == 'B'
df_B = df_Data[dfmask]

print(df_A.head(2))
print(df_F.head(2))
print(df_E.head(2))
print(df_I.head(2))
print(df_X.head(2))
print(df_H.head(2))
print(df_G.head(2))
print(df_D.head(2))
print(df_Y.head(2))
print(df_C.head(2))
print(df_W.head(2))
print(df_B.head(2))

#print(classes)
#print(N_B)'''

A    6886
F    3139
E    1725
I    1335
X     826
H     824
G     716
D     565
Y     425
C     171
W      72
B       9
Name: F11, dtype: int64
         F1       F2        F3       F4        F5        F6        F7  \
0   0.15498 -0.16562 -3.210528 -1.32068 -0.185173 -0.307984  0.484429   
2  0.080916  0.10132   0.10404  0.14049  0.261718  0.647088  0.257927   

         F8        F9       F10 F11  
0  0.340506  0.157628 -0.415334   A  
2 -0.327849 -0.247731  0.564989   A  
         F1        F2        F3        F4        F5        F6        F7  \
4  0.142636 -0.000745  0.153829  0.955845  0.261718  0.024215 -0.119577   
8 -0.042522 -0.322644  0.310311  0.287293  0.261718   0.06574  0.333428   

         F8        F9       F10 F11  
4  0.312063 -0.247731  0.209808   F  
8  0.205842 -0.123005  0.005293   F  
         F1        F2        F3        F4        F5        F6        F7  \
5 -0.425181 -0.220579  0.306755  0.496637  0.261718 -0.972381 -0.006326   
6  0.043885  0.007106  1.348781 

#Transformando os labels em números

In [ ]:
from sklearn.preprocessing import LabelEncoder
labelEncoder = LabelEncoder()
labelsLetters = df_Data['F11']

labelsClasses = labelEncoder.fit_transform(labelsLetters)

print(labelsClasses[1])
print(labelsLetters[1])

df_Data['Labels'] = labelsClasses

df_Data.head(3)

newDataSetJoined = df_Data.drop(labels=['F11'], axis = 1, inplace=False)

newDataSetJoined.head(3)


6
G


F1       F2        F3        F4        F5        F6        F7  \
0   0.15498 -0.16562 -3.210528  -1.32068 -0.185173 -0.307984  0.484429   
1  0.327793  0.50173  0.050694 -1.018301  0.261718 -0.764757  0.333428   
2  0.080916  0.10132   0.10404   0.14049  0.261718  0.647088  0.257927   

         F8        F9       F10  Labels  
0  0.340506  0.157628 -0.415334       0  
1  0.754385  0.812439  -0.74098       6  
2 -0.327849 -0.247731  0.564989       0

# Vetor de médias

---



In [ ]:
 #Calculando o vetor de médias para cada classe
def calcMedClasses(df):
  # Classe
  m = np.zeros([1, df.shape[1] - 1])
  vector = df.values

  for c in range(0, df.shape[1] - 1):
    sum = 0;
    for l in range(0, len(df)):
      sum += vector[l][c]
    m[0][c] = sum / len(df)

  #print("\n", m)

  return m

#Classe a
media_A = calcMedClasses(df_A)
media_F = calcMedClasses(df_F)
media_E = calcMedClasses(df_E)
media_I = calcMedClasses(df_I)
media_X = calcMedClasses(df_X)
media_H = calcMedClasses(df_H)
media_G = calcMedClasses(df_G)
media_D = calcMedClasses(df_D)
media_Y = calcMedClasses(df_Y)
media_C = calcMedClasses(df_C)
media_W = calcMedClasses(df_W)
media_B = calcMedClasses(df_B)

medias = [media_A, media_F, media_E, media_I, media_X, media_H, media_G, media_D, media_Y, media_C, media_W, media_B]
print(medias)

[array([[-0.0202005 ,  0.01440285, -0.10033515, -0.001378  , -0.16722192,
         0.22226106,  0.11939734, -0.30827088, -0.30869756,  0.30248587]]), array([[ 0.0806884 ,  0.03097947,  0.03565124, -0.22942204,  0.05337774,
        -0.12013641,  0.11408497, -0.263208  , -0.26209476,  0.01500051]]), array([[-0.25886491, -0.06019597,  0.12641111,  0.02222732, -0.06263406,
        -0.0602068 ,  0.19954698,  0.58174968,  0.27518996,  1.62631333]]), array([[ 0.64950097, -0.29666741, -0.08396211, -0.11542882,  0.69950337,
        -0.7550005 , -1.14640732, -0.00810636,  0.71261177,  0.0343572 ]]), array([[-0.34025369, -0.24834286, -0.03543062,  1.16516851,  0.05342142,
         0.56206061, -0.11770334,  3.90134549,  1.37983374,  1.55948325]]), array([[ 0.2469292 ,  0.12117656,  0.29065167, -0.46375071,  0.18850163,
        -0.56291136,  0.17776845, -0.09941374,  0.17147833, -1.88447164]]), array([[ 0.03376505,  0.47952543,  0.18017457, -0.20481933,  0.12103476,
        -0.16152553,  0.29302362

#Matriz de Probabilidades

In [ ]:
def p_W1(df, n):
  return n / len(df)

P_Wi = np.zeros([1, len(frequencias)])

for i in range(0, len(frequencias)):
  P_Wi[0][i] = p_W1(df_Data, frequencias[i])

print(P_Wi)